runtime/vm/resolvesupport.cpp

```diff
@@ -2105,56 +2260,80 @@ resolveInvokeDynamic(J9VMThread *vmThread, J9ConstantPool *ramCP, UDATA callSite
 {
 	Assert_VM_true(J9_RESOLVE_FLAG_RUNTIME_RESOLVE == resolveFlags);
 	j9object_t *callSite = ramCP->ramClass->callSites + callSiteIndex;
-	j9object_t methodHandle = *callSite;
+	j9object_t result = *callSite;
 
 	J9ROMClass *romClass = ramCP->ramClass->romClass;
 	J9SRP *callSiteData = (J9SRP *) J9ROMCLASS_CALLSITEDATA(romClass);
 	U_16 *bsmIndices = (U_16 *) (callSiteData + romClass->callSiteCount);
 	U_16 *bsmData = bsmIndices + romClass->callSiteCount;
 	J9ROMNameAndSignature* nameAndSig = SRP_PTR_GET(callSiteData + callSiteIndex, J9ROMNameAndSignature*);
 	U_16 bsmIndex = bsmIndices[callSiteIndex];
-	U_16 i;
+	U_16 i = 0;
 
 	/* Check if already resolved */
-	if (methodHandle != NULL) {
-		return methodHandle;
+	if (result != NULL) {
+		return result;
 	}
 
+	Trc_VM_resolveInvokeDynamic_Entry(vmThread, callSiteIndex, bsmIndex, resolveFlags);
+
 	/* Walk bsmData - skip all bootstrap methods before bsmIndex */
 	for (i = 0; i < bsmIndex; i++) {
 		/* increment by size of bsm data plus header */
 		bsmData += bsmData[1] + 2;
 	}
 
 	sendResolveInvokeDynamic(vmThread, ramCP, callSiteIndex, nameAndSig, bsmData);
-	methodHandle = (j9object_t) vmThread->returnValue;
+	result = (j9object_t) vmThread->returnValue;
+
+	Trc_VM_resolveInvokeDynamic_Resolved(vmThread, callSiteIndex, result);
 
+#if defined(J9VM_OPT_OPENJDK_METHODHANDLE)
 	/* check if an exception is already pending */
 	if (vmThread->currentException != NULL) {
 		/* Already a pending exception */
-		methodHandle = NULL;
-	} else if (methodHandle == NULL) {
+		result = vmThread->currentException;
+	} else if (result == NULL) {
+		setCurrentExceptionUTF(vmThread, J9VMCONSTANTPOOL_JAVALANGNULLPOINTEREXCEPTION, NULL);
+	}
+
+	if (result != NULL) {
+		/* store result */
+		J9MemoryManagerFunctions *gcFuncs = vmThread->javaVM->memoryManagerFunctions;
+		J9Class *j9class = J9_CLASS_FROM_CP(ramCP);
+		if (0 == gcFuncs->j9gc_objaccess_staticCompareAndSwapObject(vmThread, j9class, callSite, NULL, result)) {
+			/* Another thread beat this thread to updating the call site, ensure both threads return the same method handle. */
+			result = *callSite;
+		}
+	}
+#else /* defined(J9VM_OPT_OPENJDK_METHODHANDLE) */
+	/* check if an exception is already pending */
+	if (vmThread->currentException != NULL) {
+		/* Already a pending exception */
+		result = NULL;
+	} else if (result == NULL) {
 		setCurrentExceptionUTF(vmThread, J9VMCONSTANTPOOL_JAVALANGNULLPOINTEREXCEPTION, NULL);
 	}
```

Comment:
The below `if (result != NULL)` can be replaced with an `else` block here.
```suggestion
	} else { // (result != NULL)
```